File: leetcode/1710.卡车上的最大单元数.cpp
class Solution {
public:
    int maximumUnits(vector<vector<int>>& boxTypes, int truckSize) {
        const int n = boxTypes.size();
        int sum = 0;
        int l = 0, r = n;
        const auto ptr = boxTypes.data();
        const auto cmp = [&] (const vector<int>& box1, const vector<int>& box2) {
            return box1[1] > box2[1];
        };
        while (l < r) {
            const int mid = (l + r) / 2;
            nth_element(ptr + l, ptr + mid, ptr + r, cmp);
            int nsum = sum;
            for (int i = l;i < mid;++i)
                nsum += ptr[i][0];
            if (truckSize <= nsum)
                r = mid;
            else {
                sum = nsum + ptr[mid][0];
                l = mid + 1;
            }
        }
        int ans = 0;
        for (int i = 0;i < l;++i)
            ans += ptr[i][0] * ptr[i][1];
        if (sum > truckSize) ans -= (sum - truckSize) * ptr[l - 1][1];
        return ans;
    }
};

File: leetcode/1710.卡车上的最大单元数.py
class Solution:
    def maximumUnits(self, boxTypes: List[List[int]], truckSize: int) -> int:
        # 背包问题
        # dp[i][j]=max(dp[i-1][j],dp[i-1][j-w[i]]+p[i])
        n=len(boxTypes)
        boxTypes.sort(key=lambda x:x[1],reverse=True)
        cnt=0
        i=0
        res=0
        while i<n :
            if cnt+boxTypes[i][0]<=truckSize:
                cnt+=boxTypes[i][0]
                res+=boxTypes[i][1]*boxTypes[i][0]
            else:
                res+=(truckSize-cnt)*boxTypes[i][1]
                return res
            i+=1
        return res
                    